emulator.py

```diff
@@ -708,6 +708,172 @@ def object_acl_delete(bucket_name, object_name, entity):
+                    upload.media,
+                    upload.bucket,
+                    False,
+                    None,
+                )
+                blob.metadata.metadata["x_emulator_transfer_encoding"] = ":".join(
+                    upload.transfer
+                )
+                db.insert_object(upload.request, bucket_name, blob, None)
+                projection = testbench.common.extract_projection(
+                    upload.request, CommonEnums.Projection.NO_ACL, None
+                )
+                fields = upload.request.args.get("fields", None)
+                return testbench.common.filter_response_rest(
+                    blob.rest_metadata(), projection, fields
+                )
+            return upload.resumable_status_rest()
+        _, chunk_last_byte = [v for v in items[0].split("-")]
+        x_upload_content_length = int(
+            upload.request.headers.get("x-upload-content-length", 0)
+        )
+        if chunk_last_byte == "*":
+            x_upload_content_length = len(upload.media)
+            chunk_last_byte = len(upload.media) - 1
+        else:
+            chunk_last_byte = int(chunk_last_byte)
+        total_object_size = (
+            int(items[1]) if items[1] != "*" else x_upload_content_length
+        )
+        if (
+            x_upload_content_length != 0
+            and x_upload_content_length != total_object_size
+        ):
+            testbench.error.mismatch(
+                "X-Upload-Content-Length",
+                x_upload_content_length,
+                total_object_size,
+                None,
+                rest_code=400,
+            )
+        upload.media += data
+        upload.complete = total_object_size == len(upload.media) or (
+            chunk_last_byte + 1 == total_object_size
+        )
+    else:
+        upload.media += data
+        upload.complete = True
+    if upload.complete:
+        blob, _ = gcs_type.object.Object.init(
+            upload.request,
+            upload.metadata,
+            upload.media,
+            upload.bucket,
+            False,
+            None,
+            upload.rest_only,
+        )
+        blob.metadata.metadata["x_emulator_transfer_encoding"] = ":".join(
+            upload.transfer
+        )
+        blob.metadata.metadata["x_emulator_upload"] = "resumable"
+        blob.metadata.metadata["x_emulator_custom_header"] = str(custom_header_value)
+        db.insert_object(upload.request, bucket_name, blob, None)
+        projection = testbench.common.extract_projection(
+            upload.request, CommonEnums.Projection.NO_ACL, None
+        )
+        fields = upload.request.args.get("fields", None)
+        return testbench.common.filter_response_rest(
+            blob.rest_metadata(), projection, fields
+        )
+    else:
+        return upload.resumable_status_rest()
+
+
+@upload.route("/b/<bucket_name>/o", methods=["DELETE"])
+@retry_test(method="storage.objects.delete")
```

Comment:
It seems like we only need the retry_test route on `object_delete()`. `delete_resumable_upload()` is doing something specific and as you noted always returns code 499

Reply:
Good questions, frankly I have not looked at this stuff in detail, just trying to get it to work. But let me take a stab:

> We have the retry_test route both here and on Line#465object_delete(). 

yes.

> How does the flask app handle duplicate routes?

These are different flask applications, and different routes, `object_delete()` is for the `gcs` app, and the route is `/storage/v1/b/<bucket>/o/<object>` while this is `/**upload**/storage/v1/b/<bucket>/o` <-- note, no object path.

So probably no duplicate routes, but there are duplicate method names.  I think that is harmless.

> It seems like we only need the retry_test route on object_delete().

Not sure because:

> delete_resumable_upload() is doing something specific and as you noted always returns code 499

Presumably we want to retry upload id deletes (if supported). They are idempotent (they only succeed once) and we may want to return 429 or 503 to force a retry.

In any case: my excuse is that I am just [*] moving code.  I also happen to think this is fine, but I can delete the retry entry if you prefer.  Won't matter much either way methinks.


[*]: yes, I am also fixing bugs discovered by the (new) unit tests.

Reply:
Makes sense! We can leave it as is. That sounds like something we'd want to test retries with (if supported), and it's probably the duplicate method name that we can differentiate later on. Thanks for the clarification!